### ISSUE TITLE
Can Spotbugs enforce private visibility and single-read access for variables of a certain type?

### ISSUE THREAD

**Issue by smillies** [spotbugs/discuss#54]
I'd like to enforce best practices for the use of the SafeDCLFactory from [Alex Shipilev’s paper]( http://shipilev.net/blog/2014/safe-public-construction) on safe public construction.

In order to do that, I’d like to have Spotbugs (or Findbugs, for our legacy codelines) issue a bug report when one of the following happens:

- a class member of type `SafeDCLFactory` is not private and final
- there is more than one read access to some member of type `SafeDCLFactory`

Is that possible? If so, how? 

PS: I also think best practices are more of a style thing than  a bug thing, but none of the style checkers such as CheckStyle that I have looked at could do this analysis.

**Comment by mebigfatguy** [comment]
You'd have to write your own plugin of course, but should be relatively 
trivial.

**Comment by smillies** [comment]
I feared as much :-( Trivial, eh? Is there a good tutorial, documentation on how to go about it? The official website just has a pointer to the Javadocs, and that's a bit thin for a complete beginner like myself.

**Comment by mebigfatguy** [comment]
Guides for making detectors:

    https://www.ibm.com/developerworks/library/j-findbug2/
    http://kjlubick.github.io/blog/post/1?building-my-first-findbugs-detector

Java bytecode

    https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html

**Comment by smillies** [comment]
Thank you, that has been very helpful.

It almost works. I am still struggling with one thing, and I cannot find BCEL documentation that helps. Perhaps you can point me in the right direction. I am looking at invocations of the GET_INSTANCE_METHOD like this:

```
@Override
public void sawOpcode(int seen) {
    if (seen == Const.INVOKEVIRTUAL) {
        FQMethod fqm = new FQMethod(getClassConstantOperand(), getNameConstantOperand(), getSigConstantOperand());
        if (GET_INSTANCE_METHOD.equals(fqm)) {
            ...
        }
    }
}
```

What I need to do is to partition the method calls by the variables (fields) on which they occur. Do you know how I could do that? (or read up on it?)

**Comment by mebigfatguy** [comment]
i assume you are looking at fb-contrib as an example which is fine, if that's the case, take a look at uses of OpcodeStack. That basically holds all the things that get pushed onto the stack as the code is 'executed'. So when you get to a invokevirtual, the opcode stack will have all the parameters and also the 'this' object on which the object was called. So look at the signature of the method, get the number of args, and go down the stack that many to get the OpcodeStack.Item that is the 'this' object. You will be able to get the field name of that object form the item.

**Comment by smillies** [comment]
Thanks again, that worked out quite well! (Except for local variables, but that is a use case I don't really need to cover.)

I have a couple of questions about classpath and packaging, though. The User Guide is silent on this.

1. When running the JUnit tests for my custom detector inside Eclipse, how do I set the auxClasspath to avoid warnings about missing application classes? 
2. As you have noticed, the plugin depends on external libraries, such as fb-contrib. How do I pull those in at runtime, e. g. when calling Spobugs through Ant or in the IDE plugin (Eclipse/IntelliJ)? I would have expected the Maven archetype to conatin a suitable configuration to create a jar with such classpath entries in the manifest, or some such. I'm not really familar with this. What I finally did was use the Maven Shade plugin to create an uber.jar, but that blows up my detector plugin to about 6 MB.

**Comment by smillies** [comment]
Thanks for the info. As for using a project file, I guess not. I just followed the [Spotbugs manual](https://spotbugs.readthedocs.io/en/stable/implement-plugin.html) to create a project, imported that project into Eclipse and ran the JUnit test with the standard run configuration created by Eclipse.

**Comment by mebigfatguy** [comment]
A long time ago i recommended to the defunct findbugs project that FindBugs read the Class-Path: attribute in the manifest of the plugin (if it exists) to find aux jars needed by the plugin. I don't think it ever was implemented. Now that i think of it, probably would be good for SpotBugs to do.

**Comment by smillies** [comment]
Yes, that would be a good thing. For now, I've decided to get rid of the fb-contrib dependency by simply copying the FQMethod code to my detector. The missing auxClasspath for my application files in the Eclipse runtime config is only a very minor nuisance, because I'll always be using spotbugs from Ant.

Thanks for your support!

**Comment by smillies** [comment]
Oops, I've downloaded newest fb-contrib-7.4.3.jar and am using it with Spotbugs 3.1.6.
 I'm getting stacktraces about incompatible BCEL versions:

>  [spotbugs]   Exception analyzing com.idsscheer.ppm.server.ws.impl.ZWSServerFavoriteResolver using detector com.mebigfatguy.fbcontrib.detect.OverlyPermissiveMethod
 [spotbugs]     java.lang.RuntimeException: Incompatible bcel version, apparently bcel has been upgraded to not use 'Unknown' for 'BootstrapMethods', but uses: BootstrapMethods
 [spotbugs]       At com.mebigfatguy.fbcontrib.detect.OverlyPermissiveMethod.getBootstrapMethod(OverlyPermissiveMethod.java:451)
 ...
 [spotbugs]   Exception analyzing com.idsscheer.ppm.server.ws.impl.ZWSServerFavoriteResolver using detector com.mebigfatguy.fbcontrib.detect.FunctionalInterfaceIssues
 [spotbugs]     java.lang.ClassCastException: org.apache.bcel.classfile.BootstrapMethods cannot be cast to org.apache.bcel.classfile.Unknown
 [spotbugs]       At com.mebigfatguy.fbcontrib.detect.FunctionalInterfaceIssues.getMethodHandle(FunctionalInterfaceIssues.java:345)
 ...

What is the latest pair of compatible versions? Is there a compatibility matrix somewhere?

**Comment by mebigfatguy** [comment]
Use a version with .sb in it, like 7.4.3.sb

**Comment by smillies** [comment]
Thanks. I have another feature idea: It seems that it is possible to omit visitors only based on their class names, at least in the Ant task, which is impractical. I only want correctness warnings, and would like to exclude visitors by category. Putting the exclusion in the filter file does not prevent those zillions of unwanted detectors from running, which significantly adds to the cost of a check (my code base has >10.000 classes).
